### PR TITLE
Add create release TC server-side

### PR DESCRIPTION
**Pull request by long-sean**
- Add frontend edit and view screens
- Mandatory param validation
- Test coverage
- Agent skeleton awaiting SDK

### CODE REVIEW

File: octopus-common/src/main/java/octopus/teamcity/common/createrelease/CreateReleaseUserData.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+public class CreateReleaseUserData extends BaseUserData {
+
+  private static final CreateReleasePropertyNames KEYS = new CreateReleasePropertyNames();
+
+  public CreateReleaseUserData(Map<String, String> params) {
+    super(params);
+  }
+
+  public String getProjectName() {
```
Comment (rain-on) [inline octopus-common/src/main/java/octopus/teamcity/common/createrelease/CreateReleaseUserData.java:17]: are any of these optional?

File: octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            <span class="error" id="error_${keys.channelNamePropertyName}"></span>
+            <span class="smallNote">The channel name to release to</span>
+        </td>
+    </tr>
+    <tr>
+        <th>Packages:<l:star/></th>
+        <td>
+            <props:multilineProperty name="${keys.packagesPropertyName}" rows="5" cols="55" expanded="true" linkTitle="Enter list of package strings"/>
+            <span class="error" id="error_${keys.packagesPropertyName}"></span>
+            <span class="smallNote">Newline separated list of packages in the format - <em>PackageID:Version</em> or <em>StepName:PackageName:Version</em>.</span>
```
Comment (rain-on) [inline octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp:50]: The either/or mentality with this is a bit nasty, but I don't have a better way of dealing with it - i.e. could a user have some lines with 3 items and some with 2?

File: octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Test
+  public void describePropertiesProducesExpectedOutput() {
+    String message = step.describeParameters(buildPropertiesMap());
+    assertEquals("Project name: Project-1\nPackage version: 1.0.0", message);
+  }
+
+  @Test
+  public void validatePropertiesReturnsEmptyInvalidPropertiesListWithValidProperties() {
+    List<InvalidProperty> invalidProperties = step.validateProperties(buildPropertiesMap());
```
Comment (rain-on) [inline octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java:26]: its a bit annoying, but if possible, its worth trying to use 'final' everywhere you can - params and variables
(surprised errorprone didn't pick this up :thinking:

We may not have configured it to do so ...

File: octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Test
+  public void validatePropertiesReturnsSingleInvalidPropertyNullProjectName() {
+    Map<String, String> properties = buildPropertiesMap();
+    properties.put(CreateReleasePropertyNames.PROJECT_NAME, null);
+
+    List<InvalidProperty> invalidProperties = step.validateProperties(properties);
+    assertNotNull(invalidProperties);
+    assertEquals(1, invalidProperties.size());
+    assertEquals(
```
Comment (rain-on) [inline octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java:39]: nit: its probably worth validating the invalidProperty _field_ is set to CreateReleaePropertyNames.PROJECT_NAME

File: octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    <tr>
+        <th>Project name:<l:star/></th>a
+        <td>
+            <props:textProperty name="${keys.projectNamePropertyName}" />
+            <span class="error" id="error_${keys.projectNamePropertyName}"></span>
+            <span class="smallNote">The project's name</span>
+        </td>
+    </tr>
+    <tr>
+        <th>Package version:<l:star/></th>a
```
[the remark this answers is not in the document]
Comment (long-sean) [reply octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp:22]: Copy paste mistake

File: octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java (hunk 3 above)
Comment (long-sean) [reply octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java:26]: Done

File: octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  public void describePropertiesProducesExpectedOutput() {
+    String message = step.describeParameters(buildPropertiesMap());
+    assertEquals("Project name: Project-1\nPackage version: 1.0.0", message);
+  }
+
+  @Test
+  public void validatePropertiesReturnsEmptyInvalidPropertiesListWithValidProperties() {
+    List<InvalidProperty> invalidProperties = step.validateProperties(buildPropertiesMap());
+    assertNotNull(invalidProperties);
+    assertEquals(0, invalidProperties.size());
```
[the remark this answers is not in the document]
Comment (long-sean) [reply octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java:28]: Converted over

File: octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java (hunk 4 above)
Comment (long-sean) [reply octopus-server/src/test/java/octopus/teamcity/server/generic/CreateReleaseStepTest.java:39]: Good idea, have updated tests to check field

File: octopus-common/src/main/java/octopus/teamcity/common/createrelease/CreateReleaseUserData.java (hunk 1 above)
Comment (long-sean) [reply octopus-common/src/main/java/octopus/teamcity/common/createrelease/CreateReleaseUserData.java:17]: Updated with Optional<>

**Review by rain-on — COMMENTED** [review]
LGTM

**Review by rain-on — APPROVED** [review]

File: octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp (hunk 2 above)
Comment (rain-on) [reply octopus-server/src/main/resources/buildServerResources/v2/subpages/editCreateReleaseParameters.jsp]: :/ yuck problem - but not really an issue in TC

**Pull request merged**